### ISSUE TITLE
In-Foundry Inky editor

### ISSUE THREAD

**Issue by anthonyronda** [AaronLip/foundry-ink#1]
Dreaming big! I'm hoping to learn lessons from the Macronomicon project in this arena once that goes live

This is a three-pronged challenge
- [ ] Edit, save, and load ink source
- [ ] Story preview
- [ ] Compile ink to json, save and load json

**Comment by AaronLip** [comment]
Hey Anthony, thanks for joining the journey!

You probably know this part, and I'm including it to aid future contributors. There are a few parts in the ink ecosystem:

* The compiler which takes `.ink` source files and produces `.json`. inklecate is the official one. The [september compiler](https://github.com/thesephist/september) is on my radar, but it's not for this scripting language.
* Editors, which provide various workflow features for `.ink` source. Currently inky serves this role using inklecate as its backend.
* Interpreters, which include the official ink-engine written as a C# `.dll`, as well as inkjs which this project leverages.
* Frontends, which handle presentation logic for the interpreter.

Compiler support is not on my own roadmap yet. However, I would love to include compilation support at some time in the future!

Most of my own contributions to foundry-ink will involve polishing the frontend interfaces and making more of foundry available to writers in ink. I think the project is making progress towards being a better frontend, and I've certainly had fun now that external functions and macros can be called in stories.

[A sneak peek!](https://imgur.com/a/oGGcGhi)

But there's still so many design elements left to solidify. Things like:

* Which hooks should be provided? (I have a few ideas here which will show up soon)
* How should user permissions be integrated?
* Would the macro interface, item interface, journal interface, or something else be more intuitive as a "story access panel" for users?

Hooks in particular will be an important way that this project can interact with others in the interactive fiction sphere of modules (which includes ViNo and Theatre).

I wholeheartedly join you in dreaming big. To get there, foundry-ink needs a solid foundation. So here's to dreaming of the day when foundry-ink is a stable frontend and ready to tackle compiler integration! 🍻 The more the merrier.

Foundry-ink is in prototyping stages, so many upcoming commits on my part will reflect getting features and integration online in order to explore and test them. I'm especially aiming to perfect the stop-and-resume nature of foundry, since many ink frontends assume a single copy of the interpreter will run for the entire duration of the story's presentation to the user.

If you are taking on the challenge of a compiler + editor, feel free to reach out to me here or via discord (Animu36#5959) should you need any advice or API changes.

In summary:

Thank you, welcome, and apologies for the mess!
